Implement token exchange from OCM

enriquepablo opened the pull request:
See #57152 #57166

# Summary

This PR implements the token exchange flow from OCM, allowing Nextcloud to use bearer auth with short lived tokens. This opens up interoperability with OpenCloud/ownCloud OCIS and the possibility to implement webapp shares in the future, which requires this.

# Discussion

This is a preliminary PR for discussion. If this goes through there are some things missing with which we'll be grateful to get some guidance:

- docs and tests
- keeping retrieved access tokens in the db
- removing tokens when removing shares

There is also the question that for access tokens, we are keeping a reference to the corresponding refresh token in the uid field of the access token db record.

enriquepablo committed:
feat: Bearer auth aware Sabre HTTP client

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(dav): Add token endpoint to exchange refresh tokens for access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(dav): Add Bearer auth backend for webdav requests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(dav): New method doTryTokenLogin to allow to try token login with token rather that request

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(federatedfilesharing): Create permanent refresh token when creating a remote share

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(federatedfilesharing): When a remote requests a share with a token, it may be an access token

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(files_sharing): When requesting a remote share with bearer auth, get an access token to use as bearer token

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat: adapt to guzzle api

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(cloud_federation_api): adapt to new format for share creation

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(cloud_federation_api): support multi protocol for share creation

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): data sent to token endpoint must be application/x-www-form-urlencoded

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: federated share provider tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: share manager test

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): when receiving a share, account for the must-exchange-token requirement, in addition to the exchaange-token capability

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): POSTs to token endpoint should be signed

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): POSTs to token endpoint MUST be signed

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): fix federated share provider tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): fixing federated share provider tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): fixing federated share provider tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fixing code style

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fixing openapi specs

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix psalm issues

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: reorder import

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): do not import from NCU ns

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix sqlite integration tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): order of imports

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix session tests using Session::loginWithToken

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix public key token provider test

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: Fixed undefined $request variable

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_external): Added missing doTryTokenLogin() method to implement the IUserSession interface

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: Fixed parent::getType() to use ->getter('type') to avoid Psalm magic method detection issue

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: Added getTokenEndPoint() and setTokenEndPoint() methods that should have been moved from ICapabilityAwareOCMProvider

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix session tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(federatedfilesharing): remove unused import

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: fix user session tests

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test: test token controller

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test: test doTryTokenLogin method

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): remove unused import in TokenController test

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(dav): refresh expired tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): refactor refreshing access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): keep refresh tokens in its own db table

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): validate token exchange response before using it

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): keep refresh tokens in its own db table, add migration for old shares

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): keep access tokens in its own field in the external shares table

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): prevent concurrent requests to refresh token independently

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(dav): cleanup expired access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): prevent infinite loop trying unsuccessfully to refresh access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): missing autoloads

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test(dav): test bearer token login

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test(dav): test cleanup of expired access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test(federatedfilesharing): test federated shares

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
test(files_sharing): test access tokens

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix(files_sharing): correct access level for appConfig

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: backwards compatibility for shares from instances before upgrading

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: backwards compatibility for shares for instances before upgrading

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: backwards compatibility for shares for instances before upgrading

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: re-order imports

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix: avoid changing the IUserSession interface

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix[dav]: composer autoload

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
fix[federatedfilesharing]: replace deprecated import

Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

mickenordin committed:
chore: bump 3rdparty and update psalm baseline

- Align 3rdparty submodule with master
- Add baseline entry for the extra getFederationIdFromSharedSecret call
  introduced for refresh-token flow in RequestHandlerController

Signed-off-by: Micke Nordin <kano@sunet.se>

mickenordin commented:
Looks like all tests pass now. The filedrop test on sqlite seems very flakey, as in an artefact gets removed between it is created and accessed sometimes, but doesn't look like a problem with this code.

@ArtificialOwl @sorbaugh @icewind1991 @provokateurin would some of you mind reviewing the code?

mickenordin commented:
<img width="968" height="1075" alt="2026-04-26T18:55:39,559358562+02:00" src="https://github.com/user-attachments/assets/804ae539-2163-43f2-8ac5-07478013855f" />
Proof of tests passing if we need to push some more :stuck_out_tongue:

susnux commented on lib/private/User/Session.php:
```diff
@@ -617,14 +617,35 @@ private function loginWithToken($token) {
 
+		$this->manager->emit('\OC\User', 'preLogin', [$dbToken->getLoginName(), $password]);
+
+		// See line 173 in this module, needed for completeLogin
```
This is not really helpful the code will change in the future, better directly add a comment why this is needed rather than link to some line that will change anyways in the future.

susnux commented on apps/files_sharing/lib/Migration/Version33000Date20260306120000.php:
This sounds not like this should be implemented in `files_sharing` but rather `federatedfilessharing` app.

susnux commented on apps/cloud_federation_api/lib/Db/OcmTokenMap.php:
This looks like it should belong to the federation app itself not the dav app.

susnux replied:
should be in cloud_federation_api app

susnux reviewed (commented):
IIRC OCM is `federatedfilessharing` and `federation` so better implement it there and use proper manager / abstraction rather than span the implementation across four different apps.

Not sure how to review all of the core things, this all affects security related parts of the code. So it would be much better to split the PR into multiple minimal PRs that can be properly reviewed for security rather than one large PR 👀

mickenordin commented on apps/dav/appinfo/v1/publicwebdav.php:
```diff
@@ -90,8 +99,11 @@ function (\Sabre\DAV\Server $server) use (
-
-		$share = $authBackend->getShare();
+		try {
+			$share = $authBackend->getShare();
```
Should discover token-exchange capability

enriquepablo replied:
I think the kind of a share might not reflect the current capabilities, it may be older, no?

mickenordin commented on apps/dav/appinfo/v1/publicwebdav.php:
```diff
@@ -9,6 +9,7 @@
 use OC\Files\Storage\Wrapper\DirPermissionsMask;
 use OC\Files\View;
```
This code should go into version 2 of the dav api

mickenordin replied:
And into the four other v1 endpoints

mickenordin commented on apps/cloud_federation_api/lib/Controller/TokenController.php:
```diff
@@ -0,0 +1,191 @@
+<?php
```
This should go in the cloud_federation_api

mickenordin commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -490,6 +517,12 @@ private function confirmNotificationIdentity(
 			$provider = $this->cloudFederationProviderManager->getCloudFederationProvider($resourceType);
 			if ($provider instanceof ISignedCloudFederationProvider || $provider instanceof \NCU\Federation\ISignedCloudFederationProvider) {
 				$identity = $provider->getFederationIdFromSharedSecret($sharedSecret, $notification);
+				if ($identity === '') {
```
Try to `know` which type of auth it is, instead of guessing, maybe with token-exchange capability

enriquepablo replied:
The request we are receiving may be for a share that does not correspond with the remote's capabilities, for example because it is old

enriquepablo commented on apps/dav/appinfo/v1/publicwebdav.php:
```diff
@@ -80,6 +88,7 @@
 	$authPlugin,
 	function (\Sabre\DAV\Server $server) use (
 		$authBackend,
+		$bearerAuthBackend,
```
Do this for all v2 endpoints - future compat

enriquepablo commented on apps/dav/lib/Connector/Sabre/BearerAuth.php:
```diff
@@ -41,17 +45,27 @@ private function setupUserFs($userId) {
-		if ($this->userSession->isLoggedIn()) {
+		if ($loggedIn) {
 			return $this->setupUserFs($this->userSession->getUser()->getUID());
 		}
```
revert

enriquepablo commented on apps/dav/lib/Connector/Sabre/BearerAuth.php:
```diff
@@ -41,17 +45,27 @@ private function setupUserFs($userId) {
+		assert($share !== null);
+		return $share;
+	}
+
```
Comment the other backend does the assertion

enriquepablo commented on apps/files_sharing/lib/External/Storage.php:
```diff
@@ -72,10 +73,22 @@ public function __construct($options) {
+		// If we have a stored access token (password), use Bearer auth regardless of discovery
+		// This handles the case where the share was created with must-exchange-token
+		if (!empty($options['password'])) {
+			$authType = \OC\Files\Storage\BearerAuthAwareSabreClient::AUTH_BEARER;
```
remove this

enriquepablo commented on apps/dav/lib/Controller/TokenController.php:
```diff
@@ -0,0 +1,256 @@
+		parse_str($body, $data);
+
+		$refreshToken = $data['code'] ?? '';
+		$grantType = $data['grant_type'] ?? '';
```
these can be parameters to the controller method

enriquepablo commented on apps/cloud_federation_api/lib/Db/OcmTokenMap.php:
```diff
@@ -0,0 +1,41 @@
+		$this->addType('refreshToken', Types::STRING);
+		$this->addType('expires', Types::INTEGER);
+	}
+}
```
put this in the cloud_federation_api app

enriquepablo commented on apps/dav/lib/Migration/Version1037Date20260306120000.php:
```diff
@@ -0,0 +1,54 @@
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+#[CreateTable(table: 'dav_ocm_token_map', description: 'Maps OCM access tokens to their originating refresh tokens')]
```
remove dav from table name

enriquepablo commented on apps/federatedfilesharing/lib/OCM/CloudFederationProviderFiles.php:
```diff
@@ -146,8 +205,8 @@ public function shareReceived(ICloudFederationShare $share): string {
 			$externalShare->setRemoteId($remoteId);
-			$externalShare->setShareToken($token);
-			$externalShare->setPassword('');
+			$externalShare->setShareToken($token);  // refresh token (sharedSecret)
```
rename to refresh token  - also in db

enriquepablo commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -566,4 +569,24 @@ public function getAcceptedShares(): array {
+			$share->setAccessToken($accessToken);
+			$share->setAccessTokenExpires($expiresAt);
+			$this->externalShareMapper->update($share);
+			$this->logger->debug('Updated access token for share', ['shareToken' => substr($shareToken, 0, 8) . '...']);
```
Use shareId instead of sharetofken in the logs

enriquepablo commented on apps/files_sharing/lib/Migration/Version33000Date20260306150000.php:
```diff
@@ -0,0 +1,40 @@
+
+		return $schema;
+	}
+}
```
merge this migration with the previous one for share _external

enriquepablo commented on build/integration/features/bootstrap/Sharing.php:
```diff
@@ -632,9 +634,10 @@ private function assertFieldIsInReturnedShare(string $field, string $contentExpe
+			// A token is either:
+			// - 15 characters from ISecureRandom::CHAR_HUMAN_READABLE (legacy), or
+			// - 32 characters from ISecureRandom::CHAR_ALPHANUMERIC (new OCM tokens)
+			Assert::assertMatchesRegularExpression('/^[a-zA-Z0-9]{15,32}$/', (string)$returnedShare->$field, "Field '$field' is not a token");
```
This matches 15 - 32 instead of 15 || 32

enriquepablo commented on lib/private/Authentication/Token/PublicKeyToken.php:
```diff
@@ -204,6 +204,11 @@ public function getRemember(): int {
+	public function getType(): int {
+		return $this->getter('type');
+	}
+
```
Check that this is really needed and not in the parent

enriquepablo commented on lib/private/Federation/CloudFederationFactory.php:
```diff
@@ -31,7 +41,52 @@ class CloudFederationFactory implements ICloudFederationFactory {
+
+			try {
+				$remoteProvider = $this->ocmDiscoveryService->discover($remoteDomain);
+				$capabilities = $remoteProvider->getCapabilities();
```
getCapabilities returns an object with getters for the capabilities

enriquepablo commented on lib/private/Files/Storage/DAV.php:
```diff
@@ -61,6 +104,8 @@ class DAV extends Common {
 	/** @var bool */
 	protected $ready;
+	/** @var string The resolved bearer token for AUTH_BEARER (access token or exchanged token) */
+	protected $bearerToken;
```
also add the type

enriquepablo commented on lib/private/Files/Storage/DAV.php:
```diff
@@ -48,7 +91,7 @@ class DAV extends Common {
 	protected $user;
-	/** @var string|null */
+	/** @var int|null */
 	protected $authType;
```
also add type

enriquepablo commented on lib/private/Files/Storage/DAV.php:
```diff
@@ -371,22 +617,30 @@ public function fopen(string $path, string $mode) {
+						if ($this->authType === BearerAuthAwareSabreClient::AUTH_BEARER) {
+							$auth = [];
+							$headers = ['Authorization' => 'Bearer ' . $this->bearerToken];
+						}
```
Use capabilities in discovery always instead of trying 1st Basic and then Bearer

enriquepablo commented on lib/private/Files/Storage/DAV.php:
```diff
@@ -185,6 +257,180 @@ protected function init(): void {
+		} finally {
+			$this->retryingAuth = false;
+		}
+	}
```
Client class that will try to get access token transparently

enriquepablo commented on lib/private/Files/Storage/DAV.php:
```diff
@@ -142,9 +202,21 @@ protected function init(): void {
+			}
+			$this->bearerToken = $userName;
+		}
+
```
rename variable to token and add comment to clarify the use of username

enriquepablo commented on lib/public/Authentication/Token/IToken.php:
```diff
@@ -130,4 +139,11 @@ public function setPassword(string $password): void;
+	 * @return int One of IToken::TEMPORARY_TOKEN, IToken::PERMANENT_TOKEN, or IToken::WIPE_TOKEN
+	 * @since 32.0.0
+	 */
+	public function getType(): int;
```
check this is needed (see class as well)

enriquepablo commented on lib/private/User/Session.php:
```diff
@@ -617,14 +617,35 @@ private function loginWithToken($token) {
+		$this->manager->emit('\OC\User', 'preLogin', [$dbToken->getLoginName(), $password]);
+
+		// If we are in incognito mode, getUser (called in completeLogin) will return null
+		OC_User::setIncognitoMode(false);
```
Check that all this stuff is needed in User Session

enriquepablo reviewed (commented):
review with Kate, Ferdinand, and Micke,

enriquepablo committed:
feat(auth): support permanent OCM refresh tokens and bearer login

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(dav): accept bearer access tokens on webdav endpoints

Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(cloud_federation_api): add token exchange endpoint issuing JWT access tokens

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(cloud_federation_api): accept new protocol envelope and delegate validation

Accept both the legacy options.sharedSecret envelope and the new
protocol[name].sharedSecret form. Preserve the original cloud ID so the
factory can discover capabilities, then reset shareWith to the local
username for user lookup.

Delegate per-protocol validation to providers via the new
IValidationAwareCloudFederationProvider interface, with split exception
handling: BadRequestException -> 400, ProviderCouldNotAddShareException
-> the exception's own HTTP status (501 fallback).

In the notification handler, fall back to looking up the refresh token
via OcmTokenMapMapper when the access token cannot identify the federation.

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(ocm): advertise exchange-token capability and token endpoint

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(federatedfilesharing): create refresh tokens and sign token exchange

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
feat(files_sharing): store and refresh OCM access tokens for external shares

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

enriquepablo committed:
chore: regenerate autoloaders and update psalm baseline

Co-authored-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Micke Nordin <kano@sunet.se>
Signed-off-by: Enrique Pérez Arnaud <enrique@cazalla.net>

mickenordin committed:
fix(cloud_federation_api): accept multi-protocol share envelopes

Shares using the OCM multi-protocol envelope (name multi, with the secret carried in a sibling protocol entry such as webdav) were rejected with Missing sharedSecret in protocol. Scan every protocol entry for the shared secret during validation, resolve the secret from the matching entry, and let the files provider serve the webdav entry of a multi envelope. Covers the file and folder resource types.

Signed-off-by: Micke Nordin <kano@sunet.se>

CarlSchwan commented on apps/cloud_federation_api/lib/Controller/TokenController.php:
```diff
@@ -0,0 +1,274 @@
+/**
+ * Controller for the /token endpoint
+ * Exchanges long-lived refresh tokens for short-lived access tokens
+ *
+ * @since 32.0.0
```
```suggestion
```

Only needed for ocp apis

nickvergessen commented on apps/federatedfilesharing/lib/Migration/Version1012Date20260306120000.php:
```diff
@@ -0,0 +1,119 @@
+
+	#[\Override]
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		$db = Server::get(IDBConnection::class);
+		$tokenProvider = Server::get(PublicKeyTokenProvider::class);
+		$userManager = Server::get(IUserManager::class);
```
Should use DI?

enriquepablo replied:
I have opened a (~~#61439~~) #61441 addressing this. I can see you took care of the rest of your comments.

nickvergessen commented on apps/federatedfilesharing/lib/Migration/Version1012Date20260306120000.php:
```diff
@@ -0,0 +1,119 @@
+ * Shares created by this fork (32-char tokens) that are somehow missing from
+ * oc_authtoken are silently repaired.
+ */
+class Version1012Date20260306120000 extends SimpleMigrationStep {
```
Should bump the version of the app since a migration was added

enriquepablo replied:
I can see you already addressed this in #61428 . Thanks!

nickvergessen commented on apps/files_sharing/lib/Migration/Version33000Date20260306120000.php:
```diff
@@ -0,0 +1,89 @@
+		}
+
+		if (!$table->hasColumn('access_token_expires')) {
+			$table->addColumn('access_token_expires', Types::INTEGER, [
```
If this is supposed to be a timestamp, the column type is wrong at least on 32bits

Should have lenght 11 or more, or BIGINT or type datetime instead
